telemetry/data_feature_usage.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/atomic"
+)
+
+type featureUsageInfo struct {
+	AsyncCommitUsed  bool                       `json:"asyncCommitUsed"`
+	CoprCacheUsed    []*CoprCacheUsedWindowItem `json:"coprCacheUsed"`
+	ClusterIndexUsed map[string]bool            `json:"clusterIndexUsed"`
+	TiFlashUsed      []*TiFlashUsageItem        `json:"tiFlashUsed"`
+}
+
+// FeatureTaskChan is the update task channel for telemetry feature data.
+var FeatureTaskChan = make(chan *FeatureTask, 1<<16)
+
+// FeatureTask is the update task for telemetry feature data.
+type FeatureTask struct {
+	TiFlashPushDown         bool
+	TiFLashExchangePushDown bool
+	CoprRespTimes           uint64
+	CoprCacheHitNum         uint64
+}
+
+// UpdateFeature update feature data for one task.
+func UpdateFeature(task *FeatureTask) {
+	CoprocessorCacheTelemetry.Lock.Lock()
+	length := len(CoprocessorCacheTelemetry.MinuteWindow)
+	if length == 0 || time.Since(*CoprocessorCacheTelemetry.MinuteWindow[length-1].BeginAt) >= time.Minute {
+		var i int
+		for i = 0; i < length && time.Since(*CoprocessorCacheTelemetry.MinuteWindow[i].BeginAt) >= ReportInterval; i++ {
+		}
+		CoprocessorCacheTelemetry.MinuteWindow = CoprocessorCacheTelemetry.MinuteWindow[i:]
```

Comment:
Will it cause problem (e.g. panic) when `i == length`?

Reply:
I will add a check here.

Reply:
Actually, if `i == length`. It's valid here.

```go
a = a[len(a):] // clear the array.
```